### DISCUSSION TITLE
Auth problem that drives me mad;  mystery cookies after Oauth redirect; 3 days NO sleep!

### DISCUSSION THREAD

**Discussion by nikolicstjepan** [supabase/supabase#6391] [General]
I'm trying to build a Remix app using Supabase. Idea is to use Supabase only on the server-side. I'm connecting to Supabase using `service_role` api key. 

I have a "problem" regarding Oauth, flow is;

1. User clicks on Google or Facebook icon
2. User gives his permission
3. User gets redirected back to my site with tokens in URL

Now I understand that if there is JS Supabase lib on client-side that it will handle user session. But that is not the case for me.

The thing that drives me mad is cookies that get set (after Oauth redirect, when I query Supabase DB).

`sb-access-token`
`sb-refresh-token`

Interesting thing is that these tokens are set with a different domain (https://dqyssuaewicguwdbdkdt.supabase.co), does that mean that they are created from the Supabase side (https://dqyssuaewicguwdbdkdt.supabase.co)?

These tokens do belong to the authorized user and that is not the problem, I just want to know how are they set (there is no Supabase JS client on the front), I want to remove them because I don't need them and I can't read them on server-side because they have another domain.

I tried to find an answer in supabase-js lib but I failed!

When I don't query supabase db on the server side on redirected page (when tokens are present in URL), cookies are not created. but If I change the code to use supabase db on that page, remove tokens from URL and reset the server - cookies magically appear!

Please help!

Edit: From the network tab, I learned that `sb-access-token` gets set when the Oath provider redirects to https://dqyssuaewicguwdbdkdt.supabase.co/auth/v1/callback route and that makes sense. 

Still not sure when the refresh token gets created and how to:
1. remove these cookies,
2. use these cookies in my backend

Edit 2:
These tokens are set by gotrue server!

**Comment by planetmall** [comment]
Curious if you were able to figure this out. I am going through this same issue any help would be much appreciated!

**Reply by nikolicstjepan** [reply]
@planetmall yes, gotrue server is setting those cookies and if you have supabase-js client available on the frontend they handle the session automatically.

**Reply by The-Code-Monkey** [reply]
How do you get those tokens to be deleted as i am running supabaseClient.auth.signOut() which isnt removing the tokens leaving the user signed in.

**Reply by syno3** [reply]
i am going through the same problem has the issue been resolved

**Reply by walkerdustin** [reply]
I currently have the same problem.

My auth is only on the clientside.
I use sveltekit (as a static site generator)
after `signOut()` there are still these cookies and `supabase.auth.getUser()` still gets me the user 
so user does not get logged out.

what helped me is to reload the page after the `signout()`
aka not use the linking and `goto()` function of my framework but rather use browser function
`window.location.href = '/loggedOut';`

**Comment by walkerdustin** [comment]
I currently have the same problem.

My auth is only on the clientside.
I use sveltekit (as a static site generator)
after `signOut()` there are still these cookies and `supabase.auth.getUser()` still gets me the user 
so user does not get logged out.

what helped me is to reload the page after the `signout()`
aka not use the linking and `goto()` function of my framework but rather use browser function
`window.location.href = '/loggedOut';`

**Comment by AliQ80** [comment]
Hi, 

I also have the same problem, I'm using Nuxt 3 with supabase nuxt module.

```js
async logout() {
      const client = useSupabaseClient()
      await client.auth.signOut()
    }
```

dose not remove the cockies
sb-access-token
sb-refresh-token

I have in composables useAuth that is invoked in app.vue

```js
export const useAuth = () => {
  const { $supabase } = useNuxtApp()
  const user = useState('user')

  $supabase.auth.onAuthStateChange(async (event, session) => {
    if (event === 'SIGNED_IN') {
      user.value = await $supabase.auth.getUser()
    }

    if (event === 'SIGNED_OUT') {
      user.value = null
    }
  })
}
```
but event SIGNED_OUT is never fired

I also have in plugins

```js
export default defineNuxtPlugin(async (nuxtApp) => {
  const supabase = useSupabaseClient()
  const {
    data: { user },
  } = await supabase.auth.getUser()
  nuxtApp.provide('supabase', supabase)
  useState('user', () => user)
})
```

what am I missing, thank you

**Reply by JustMrMendez** [reply]
SveleKit here, my app works, but this is driving me crazy, anyone has any idea?

**Reply by hf** [reply]
Hey `onAuthStateChange` is meant to run on the browser, not on the server.

**Comment by kangmingtay** [comment]
Hey @nikolicstjepan, yes these cookies are set by the gotrue server. If you want to send the session to your server-side, you need to get the session from the client-side and send the access and refresh tokens to the server via an authorization header or in the request body. 

You will have to set the cookie on your own either on the client-side or in the server-side. This will set the cookie on your domain, which you can then pass around throughout the rest of your application or to other sub-domains to read. 

The refresh token is created after the `/callback` route succeeds in the gotrue server. The gotrue server then sets the cookie on the `supabase.co` domain and also sets the query fragments (which contains the access token, refresh token, etc..). These query fragments will be parsed by the supabase-js client lib after the `/callback` route redirects to your site. 

The supabase-js client lib will save these query fragments in local storage by default if you set `persistSession` to true when you call `createClient(SUPABASE_URL, SUPABASE_KEY, { auth: { persistSession: true, ... }, ... }`

**Reply by hf** [reply]
@nikolicstjepan Even though you're not using a SSR framework, I believe you can find some more information [in this document](https://supabase.com/docs/guides/auth/server-side-rendering) about how the Auth flow works for server-side environments. At this time your app must have a small JavaScript (browser) component that will use the Subapase JavaScript library to manage the session and cookies properly for you.

The `sb-access-token` and `sb-refresh-token` cookies are there for backward compatibility and we no longer recommend using them (and as you noted, it's very difficult to use them across domains).

**Reply by nikolicstjepan** [reply]
Thanks for the response @kangmingtay  @hf . I understand it now! :)
